source/manual/architecture-deep-dive.html.md

```diff
@@ -0,0 +1,462 @@
+
+[cache-clearing-service]: https://github.com/alphagov/cache-clearing-service
+[Fastly]: https://www.fastly.com/
+[our-cdn]: /manual/cdn.html
+[Varnish]: https://varnish-cache.org/
+
+#### Failover
+
+If a Fastly request to origin returns a 5xx response, Fastly will request
+content from the mirror, which is static HTML hosted in an S3 bucket on AWS.
+The contents of the mirror are updated daily via a [govuk-crawler-worker],
+which recursively crawls GOV.UK URLs from a message queue, visiting the pages
+and saving the output to disk.
+
+[govuk-crawler-worker]: https://github.com/alphagov/govuk_crawler_worker
+
+#### Routing on the CDN
+
+As well as for caching, Varnish is used for the redirection from `gov.uk` to
+`www.gov.uk`, which is configured in Varnish Configuration Language (VCL) and
+uploaded directly to Fastly via [govuk-cdn-config].
+
+Other redirects that happen at the Fastly level include [bouncer]: a GOV.UK
+application responsible for redirecting traffic from old pre-GOV.UK websites.
+This is configured via [transition] and [transition-config].
+
+[bouncer]: https://github.com/alphagov/bouncer
+[govuk-cdn-config]: https://github.com/alphagov/govuk-cdn-config
+[transition]: https://github.com/alphagov/transition
+[transition-config]: https://github.com/alphagov/transition-config
+
+### Routing on GOV.UK
+
+#### Getting to the 'router' application
+
+![](images/load-balancer.png)
+
+Some requests make it through the CDN and cache layers to 'origin'. Origin is
+a stack of computers in the cloud - in this case, AWS - and its entry point is
+a load balancer.
+
+The load balancer knows based on the hostname which machine 'class' to route
+to. Different classes of machine run different sets of GOV.UK applications.
+How many machines are allocated to a class - and how big those machines are -
+is configured using [Terraform], in [govuk-aws]. What runs on each machine
+class is configured in govuk-puppet, a file and process management system we'll
+cover in more detail later.
+
+![](images/nginx-routing.png)
+
+External requests are routed to a 'cache' machine, which uses Varnish for its
+caching. If Varnish has the route response in its cache, it returns that,
+otherwise it forwards the request to [Nginx], a web server running on the
+machine. Nginx is configured with govuk-puppet.
+
+Nginx proxies most requests to [router], which is a GOV.UK maintained
+application running on the cache machines. However, Nginx does proxy some
+routes directly to other apps, such as image URLs routing to asset-manager.
+
+[govuk-aws]: https://github.com/alphagov/govuk-aws
+[Nginx]: https://www.nginx.com/
+[router]: https://github.com/alphagov/router
+[Terraform]: https://www.terraform.io/
+
+#### Routing via 'router'
+
+'router' is a reverse proxy app written in Go. It stores all known routes
+in memory using a [prefix trie]. The routes are loaded into the trie from
+[redis] at startup. Matched routes are proxied through to the relevant
```

Comment:
I believe the routes come from the MongoDB database shared with the Router API.

Reply:
I've added a line about this now.

Reply:
I think Redis is still mentioned as what the Router uses.

Reply:
Oh, I thought Router loaded routes into its prefix trie from Redis at startup - and that Redis was populated via Router API and the MongoDB. This always did seem unnecessarily complicated!

I see there's actually no mention of Redis anywhere in the router repository, but it is mentioned in [configuration in router-api](https://github.com/alphagov/router-api/blob/dae159a1852fa12cc3d7654af2fd53162b2bd6fa/config/cable.yml#L7-L10). I'm now not exactly sure where it fits in. It looks like that config is used by [Action Cable](https://guides.rubyonrails.org/action_cable_overview.html) for building real-time features, so Redis is potentially used as a temporary store when routes are updated at runtime? That seems like such a transitory implementation detail that it should just be omitted entirely from this overview. What do you think?

Reply:
> I see there's actually no mention of Redis anywhere in the router repository, but it is mentioned in [configuration in router-api](https://github.com/alphagov/router-api/blob/dae159a1852fa12cc3d7654af2fd53162b2bd6fa/config/cable.yml#L7-L10). I'm now not exactly sure where it fits in. It looks like that config is used by [Action Cable](https://guides.rubyonrails.org/action_cable_overview.html) for building real-time features, so Redis is potentially used as a temporary store when routes are updated at runtime? That seems like such a transitory implementation detail that it should just be omitted entirely from this overview. What do you think?

I'm pretty confident that we're not using Redis with the Router API. While it's mentioned in the Action Cable configuration, I don't think we're using Action Cable in the Router API.

Reply:
Thanks - I've now removed references to Redis.